src/connection.rs

```diff
@@ -432,7 +431,21 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
                 num_terminated += 1;
                 Either::Left(future::pending())
             } else {
-                Either::Right(self.socket.next())
+                // poll socket for next incoming frame, but also make sure any pending writes are properly flushed
+                let socket = &mut self.socket;
+                let next_frame = future::poll_fn(move |cx| {
+                    if let Poll::Ready(res) = socket.poll_next_unpin(cx) {
+                        return Poll::Ready(res);
+                    }
+
+                    if let Poll::Ready(Err(err)) = socket.poll_flush_unpin(cx) {
```

Comment:
I guess we assume that for all implementations of `Sink` `poll_flush_unpin` is cheap? If not, we could stop calling `poll_flush_unpin` once it returns `Poll::Ready(Ok(()))`.

Reply:
I think it makes sense. But we need to make sure to reset every time we write to the underlying IO